File: src/main/java/com/kasisoft/libs/common/functional/KConsumer.java
package com.kasisoft.libs.common.functional;

import com.kasisoft.libs.common.*;

import jakarta.validation.constraints.*;

import java.util.function.*;

/**
 * @author daniel.kasmeroglu@kasisoft.com
 */
@FunctionalInterface
public interface KConsumer<T> {

    void accept(T input) throws Exception;

    @NotNull
    default KConsumer<T> andThen(@NotNull KConsumer<? super T> after) {
        return $t -> {
            accept($t);
            after.accept($t);
        };
    }

    @NotNull
    default Consumer<T> protect() {
        return $t -> {
            try {
                accept($t);
            } catch (Exception ex) {
                throw KclException.wrap(ex);
            }
        };
    }

} /* ENDINTERFACE */

File: src/main/kotlin/com/kasisoft/libs/common/functional/KConsumer.kt
package com.kasisoft.libs.common.functional

import com.kasisoft.libs.common.*

import java.util.function.*

/**
 * This implementation allows to throw Exceptions (useful on the Java side)
 *
 * @author daniel.kasmeroglu@kasisoft.com
 */
fun interface KConsumer<T: Any?> {

    @Throws(Exception::class)
    fun accept(t: T)

    fun protect(): Consumer<T> =
        Consumer<T> { t: T ->
            try {
                accept(t)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

} /* ENDINTERFACE */
